### DISCUSSION TITLE
Custom type for model with relational fields breaks code completion and intellisense?

### DISCUSSION THREAD

**Discussion by manuelpoelzl** [prisma/prisma#22506] [Q&A]
### Question

I created a custom type for my model which includes relational fields like in the docs: [Docs](https://www.prisma.io/docs/orm/prisma-client/type-safety/operating-against-partial-structures-of-model-types).

This works as described and I do now have access to the fields, however this is somehow messing up the code completion and intelisense in my IDE.

Before these changes everything was working as expected, each operation was picked up by the IDE correctly:
![image](https://github.com/prisma/prisma/assets/46539758/d330a640-a755-446c-8f2e-307754686f03)

But after defining my custom type like this:
```typescript
const artifactWithRelations = Prisma.validator<Prisma.ArtifactDefaultArgs>()({
    include: {Epoch: true, AssignedSite: true}
});

type ArtifactWithRelationProps = Prisma.ArtifactGetPayload<typeof artifactWithRelations>;
```

Intellisense is somehow broken and can not auto complete field names anymore,.
This was before defining the type:
![image](https://github.com/prisma/prisma/assets/46539758/8a1161da-9615-4cec-842e-ad2af542c3cc)

As you can see `AssignedSite` is picked up perfectly fine, but after the change it looks like this:
![image](https://github.com/prisma/prisma/assets/46539758/f44d6e57-d01a-4a45-aa47-447eb1cb7278)

And if I press the `Tab` key to complete the  field name, this gets inserted instead:

```typescript
AssignedSite<T = {} extends Prisma.ExcavationSiteDefaultArgs<ExtArgs>>(args?: Prisma.Subset<T, Prisma.ExcavationSiteDefaultArgs<ExtArgs>>): Prisma.Prisma__ExcavationSiteClient<GetResult<Prisma.$ExcavationSitePayload<ExtArgs>, T, "findUniqueOrThrow"> | Null, Null, ExtArgs> {
```

Is there any way to overcome this issue?

I saw that #10928 is pretty similar and the reference to the new `satisfies` operator.
Although I can not see how this would be applicable here.

Thanks!

### How to reproduce (optional)

1. Create a custom type like in the  [Docs](https://www.prisma.io/docs/orm/prisma-client/type-safety/operating-against-partial-structures-of-model-types).
2. Perform some operations with the prisma client
3. Observer the code completion

### Information about Prisma Schema, Client Queries and Environment (optional)

Environment variables loaded from .env
prisma                  : 5.6.0                                                                                                                   
@prisma/client          : 5.6.0                                                                                                                   
Computed binaryTarget   : windows                                                                                                                 
Operating System        : win32                                                                                                                   
Architecture            : x64                                                                                                                     
Node.js                 : v20.10.0                                                                                                                
Query Engine (Node-API) : libquery-engine e95e739751f42d8ca026f6b910f5a2dc5adeaeee (at node_modules\@prisma\engines\query_engine-windows.dll.node)
Schema Engine           : schema-engine-cli e95e739751f42d8ca026f6b910f5a2dc5adeaeee (at node_modules\@prisma\engines\schema-engine-windows.exe)  
Schema Wasm             : @prisma/prisma-schema-wasm 5.6.0-32.e95e739751f42d8ca026f6b910f5a2dc5adeaeee                                            
Default Engines Hash    : e95e739751f42d8ca026f6b910f5a2dc5adeaeee                                                                                
Studio                  : 0.495.0

**Comment by ludralph** [comment]
HI @manuelpoelzl 👋 

Thank you for raising this question.

Did you try restarting the Typescript server in VSCode? Are you using a custom output path for your generated types?

**Reply by manuelpoelzl** [reply]
Hi!
Yes, I've restarted the TS server and I don't use a custom output path for the generated types.
I just tried it again in version 5.7.1 of Prisma, and I can still reproduce this behavior.

**Reply by nikolasburk** [reply]
Hey @manuelpoelzl 👋 I follow your instructions to reproduce the problem but I still have the autocomplete working. 

I did the following:

```
npx try-prisma@latest --template typescript/script -i npm -p . -n custom-types
cd custom-types
npx prisma migrate dev --name init
```

Then I put this code into `index.ts`:

```ts
import { PrismaClient, Prisma } from '@prisma/client'

const prisma = new PrismaClient()

// 1: Define a type that includes the relation to `Post`
const userWithPosts = Prisma.validator<Prisma.UserDefaultArgs>()({
  include: { posts: true },
})

// 2: Define a type that only contains a subset of the scalar fields
const userPersonalData = Prisma.validator<Prisma.UserDefaultArgs>()({
  select: { email: true, name: true },
})

// 3: This type will include a user and all their posts
type UserWithPosts = Prisma.UserGetPayload<typeof userWithPosts>
```

When I want to create a new object of type `UserWithPosts`, I do get autocompletion as expected:

<img width="386" alt="image" src="https://github.com/prisma/prisma/assets/4058327/1038861a-260b-457c-a7a2-c78dd5ca444c">

Can you maybe share a standalone reproduction of the issue so that we can look into it further?

**Reply by manuelpoelzl** [reply]
Hi @nikolasburk !

I think the problem has more to do with JetBrains IDEs and their Prisma plugin than with Prisma itself.
I just did a new VSCode installation as a test and tried the whole thing again, I can confirm that code completion works in VSCode

![image](https://github.com/prisma/prisma/assets/46539758/2a94cefd-e903-4375-aca6-a0a4df3dd0ca)

I think this discussion can be closed. I'll get in contact with JetBrains and see if they offer a solution, since it seems to only affetc their IDEs

Thanks!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
